### ISSUE TITLE
How to preserve upstream DNS options between reinstalls/upgrades?

### ISSUE THREAD

**Issue by SteveW25561** [pi-hole/pi-hole#333]
I have configured a few upstream DNS servers (eg TVunblock) that I want to use in preference to Open DNS (although I'd like to list openDNS as a 3rd and 4th fallback in case the TVunblock DNS is down).

The current installer only gives google or openDNS as options, and my customizations in /etc/dnsmasq.d/01-pihole.conf are overwritten each time I reinstall using the curl command.

I know that I can add lines to the dnsmasq.conf file and these are left alone when updating.  

My question is: what order will the DNS entries be queried if I have `server=` lines in **dnsmasq.conf**  as well as the upstream DNS server chosen in the install script (which I see is written into the 01-pihole.conf file)?

**Comment by PromoFaux** [comment]
What I personally do is have any custom settings (including TVunblock) in the file `/etc/dnsmasq.d/02-custom.conf`

seems to be working for me so far!

**Comment by PromoFaux** [comment]
```
fap@fappotron:~$ cat /etc/dnsmasq.d/02-unblock.conf
server=/netflix.com/46.101.36.120
server=/netflix.com/46.101.149.135
```

**Comment by dschaper** [comment]
Yep, just as @PromoFaux says, if you have domain specific hosts that you want to bypass the Pi-hole `dnsmasq` server, you can add in another config file with your overrides. The operational process is that the files are processed alphanumerically, so `a.conf` is processed, and then `b.conf`. What you have to be very careful of though is that you can not duplicate keywords in multiple files. `dnsmasq` will throw an error that you can see with a `sudo systemctl status dnsmasq -l` and you'll see the error state that it found duplicate keyword entries.

**Comment by SteveW25561** [comment]
Thanks for the replies!  I'm new to this, so please bear with me.  I wanted to make sure this is explicit.

If I understand correctly, if I have OpenDNS chosen as my upstream DNS server when using the curl install command, it will be written in 01-pihole.conf, and then if I set up another file like @PromoFaux suggests (/etc/dnsmasq.d/02-unblock.conf), then the following happens:
1. all DNS requests will go through OpenDNS preferentially, except:
2. netflix.com requests will route through the stated DNS servers

A related question for unblocking (I realize this isn't directly related to pi-hole): Have you found this works properly for Netflix apps installed on devices like the AppleTV, Roku, etc which have ways of internally routing to googleDNS (I have static routing already set up in DD-WRT to avoid the google DNS requests)?  I'm wondering if the server=/netflix.com/xx.xx.xx.xx is sufficient when paired with static routing setups.

**Comment by PromoFaux** [comment]
This works for me with netflix running on a Nexus Player, and at least according to the pi-hole admin page, requests are going to the unblock DNS servers for netflix

![image](https://cloud.githubusercontent.com/assets/1998970/13171911/28385c92-d6ed-11e5-8293-84f5c80d8762.png)

(You wont see this on your admin page untill we push the next update... [soon™](http://wowwiki.wikia.com/wiki/Soon)!)

**Comment by slamanna212** [comment]
Oh wow, that is cool, I have a couple of those entries in my config but i thought those were just static dns entries, Meaning, using the above example, we are telling the Pi Hole that Netflix is 46.101.36.120. Was i correct in that, or does that code mean to use 46.101.36.120 as the DNS server for Netflix?

**Comment by PromoFaux** [comment]
>or does that code mean to use 46.101.36.120 as the DNS server for Netflix

this.

**Comment by slamanna212** [comment]
Ok, so then how do i manually add DNS entries. As in, i want to say apple.com is IP address 1.1.1.1 or microsoft.com is 2.2.2.2 , is it possible to have it just use an address i set for a domain and not do a dns lookup?

**Comment by PromoFaux** [comment]
Totally possible!

[here is a comment I made on reddit recently](https://www.reddit.com/r/pihole/comments/469y55/redirect_domains_to_something_other_than_127001/d03ju9r?context=3) which explains the process for someone with a similar request!

**Comment by slamanna212** [comment]
@PromoFaux Awesome thanks! Also, yay, a subreddit

**Comment by PromoFaux** [comment]
Something along these lines:

>You're going to need to create a couple of additional files. 

>    ```sudo touch /etc/dnsmasq.d/customRedirect.conf```

>   ``` sudo touch /etc/customRedirect.list```

>The files can actually be called whatever you want, but the first file MUST be in `/etc/dnsmasq.d`

>in `/etc/dnsmasq.d/customRedirect.conf` you'll need the following:

>    addn-hosts=/etc/customRedirect.list

>and in `/etc/customRedirect.list` is where you put your blocklist. For example:

>    ```<IP address> bing.com```

>Also, if the domain you are trying to redirect elsewhere is already in pi-hole's blacklist, it is going to be duplicated there, run `whitelist.sh <domain>` so that it is ignored by pi-hole, and is instead picked up by `customRedirect.list`

**Comment by slamanna212** [comment]
@promoFaux thats cool but it isnt exactly what i wanted. But really, i just want the DC i just setup to stop asking the Pi-hole for random addresses it doesnt know so with the current setup of just forwarding those DNS querys to itself works. But thank you for that info!

**Comment by PromoFaux** [comment]
We'll better document this on the [wiki](https://github.com/pi-hole/pi-hole/wiki)

**Comment by DonLexos** [comment]
I added the redirect domain1.com to domain2.com to the FAQ wiki

**Comment by harshil2105** [comment]
This no longer works....even after editing the file I got no result and maybe I made some error in editing the last file....any help would be appreciated

**Comment by Mcat12** [comment]
Please open a new issue with more detailed information (follow the template; what have you tried, what errors are you seeing, etc).
